chain-storage/src/lib.rs

```diff
@@ -1,1079 +1,827 @@
+        block_info.add_tag_ref();
+        let info_bin = block_info.serialize()?;
+        info.insert(block_id, info_bin)?;
+    } else if !permanent_store_index
+        .get(block_id)
+        .map(|maybe_block| maybe_block.is_some())?
+    {
+        return Err(Error::BlockNotFound.into());
     }
 
-    pub fn generate_chain<R: RngCore>(
-        rng: &mut R,
-        store: &mut BlockStoreConnection<Block>,
-    ) -> Vec<Block> {
-        let mut blocks = vec![];
-
-        let genesis_block = Block::genesis(None);
-        store.put_block(&genesis_block).unwrap();
-        blocks.push(genesis_block);
-
-        for _ in 0..10 {
-            let mut parent_block = pick_from_vector(rng, &blocks).clone();
-            let r = 1 + (rng.next_u32() % 9999);
-            for _ in 0..r {
-                let block = parent_block.make_child(None);
-                store.put_block(&block).unwrap();
-                parent_block = block.clone();
-                blocks.push(block);
-            }
-        }
+    let maybe_old_block_id = tags.insert(tag_name, block_id)?;
 
-        blocks
+    if let Some(old_block_id) = maybe_old_block_id {
+        let info_bin = info
+            .get(old_block_id.clone())?
+            .ok_or(ConsistencyFailure::TaggedBlock)?;
+        let mut block_info = BlockInfo::deserialize(&info_bin[..], id_size, old_block_id.to_vec())?;
+        block_info.remove_tag_ref();
+        let info_bin = block_info.serialize()?;
+        info.insert(block_info.id().as_ref(), info_bin)?;
     }
 
-    #[test]
-    pub fn test_put_get() {
-        let mut store = BlockStoreBuilder::file("file:test_put_get?mode=memory&cache=shared")
-            .build()
-            .connect()
-            .unwrap();
-        assert!(store.get_tag("tip").unwrap().is_none());
-
-        match store.put_tag("tip", &BlockId::zero()) {
-            Err(Error::BlockNotFound) => {}
-            err => panic!(err),
-        }
+    Ok(())
+}
 
-        let genesis_block = Block::genesis(None);
-        store.put_block(&genesis_block).unwrap();
-        let (genesis_block_restored, block_info) = store.get_block(&genesis_block.id()).unwrap();
-        assert_eq!(genesis_block, genesis_block_restored);
-        assert_eq!(block_info.block_hash, genesis_block.id());
-        assert_eq!(block_info.chain_length, genesis_block.chain_length().0);
-        assert_eq!(block_info.back_links.len(), 1);
-        assert_eq!(block_info.parent_id(), BlockId::zero());
-
-        store.put_tag("tip", &genesis_block.id()).unwrap();
-        assert_eq!(store.get_tag("tip").unwrap().unwrap(), genesis_block.id());
+#[inline]
+#[allow(clippy::too_many_arguments)]
+fn remove_tip_impl(
+    blocks: &TransactionalTree,
+    info: &TransactionalTree,
+    chain_length_to_block_ids: &TransactionalTree,
+    tips: &TransactionalTree,
+    permanent_store_index: &TransactionalTree,
+    block_id: &[u8],
+    root_id: &[u8],
+    id_size: usize,
```

Comment:
Maybe instead of dragging these around everywhere, they should be in a `StoreOptions` struct.